rclrs/src/node/service.rs

```diff
@@ -0,0 +1,185 @@
+use std::borrow::Borrow;
+use std::boxed::Box;
+use std::ffi::CString;
+use std::sync::Arc;
+
+use crate::error::{RclReturnCode, ToResult};
+use crate::Node;
+use crate::{rcl_bindings::*, RclrsError};
+
+use rosidl_runtime_rs::Message;
+
+use crate::node::publisher::MessageCow;
+
+use parking_lot::{Mutex, MutexGuard};
+
+// SAFETY: The functions accessing this type, including drop(), shouldn't care about the thread
+// they are running in. Therefore, this type can be safely sent to another thread.
+unsafe impl Send for rcl_service_t {}
+
+/// Internal struct used by services.
+pub struct ServiceHandle {
+    handle: Mutex<rcl_service_t>,
+    node_handle: Arc<Mutex<rcl_node_t>>,
+}
+
+impl ServiceHandle {
+    pub(crate) fn lock(&self) -> MutexGuard<rcl_service_t> {
+        self.handle.lock()
+    }
+}
+
+impl Drop for ServiceHandle {
+    fn drop(&mut self) {
+        let handle = self.handle.get_mut();
+        let node_handle = &mut *self.node_handle.lock();
+        // SAFETY: No preconditions for this function
+        unsafe {
+            rcl_service_fini(handle as *mut _, node_handle as *mut _);
+        }
+    }
+}
+
+/// Trait to be implemented by concrete Service structs
+/// See [`Service<T>`] for an example
+pub trait ServiceBase: Send + Sync {
+    /// Internal function to get a reference to the `rcl` handle.
+    fn handle(&self) -> &ServiceHandle;
+    /// Tries to take a new request and run the callback with it.
+    fn execute(&self) -> Result<(), RclrsError>;
+}
+
+type ServiceCallback<Request, Response> =
+    Box<dyn Fn(&rmw_request_id_t, Request) -> Response + 'static + Send>;
+
+/// Main class responsible for responding to requests sent by ROS clients
+pub struct Service<T>
+where
+    T: rosidl_runtime_rs::Service,
+{
+    pub(crate) handle: Arc<ServiceHandle>,
+    /// The callback function that runs when a request was received.
+    pub callback: Mutex<ServiceCallback<T::Request, T::Response>>,
+}
+
+impl<T> Service<T>
+where
+    T: rosidl_runtime_rs::Service,
+{
+    /// Creates a new service.
+    pub fn new<F>(node: &Node, topic: &str, callback: F) -> Result<Self, RclrsError>
+    where
+        T: rosidl_runtime_rs::Service,
+        F: Fn(&rmw_request_id_t, T::Request) -> T::Response + 'static + Send,
```

Comment:
Using `Fn` here is inconsistent with the use of `FnMut` in subscriptions. I'd like to keep it `FnMut` here too, and to open a follow-up issue for the proposal to change it to `Fn`.

Reply:
Maybe, but don't you agree with the principle of doing changes like `FnMut` -> `Fn` atomically over the code base (services + subscriptions) so that it is never inconsistent across our code? I.e. I think we should open a separate issue (where we can maybe ensure that this is indeed the right thing to do – I don't want to stall this PR with this discussion) and a separate PR which switches to `Fn` (which would also involve removing the Mutex in line 61, for instance).

Reply:
I appreciate you caring about this so much, but I've already explained why I want to keep it as `Fn` for now, and that I agree with @mxgrey 's feedback (I originally had it as `FnMut`, in fact), but I don't want to block this PR. Feel free to open a new ticket and we can revisit this in the future. Thanks.

Reply:
I've created a follow-up issue: https://github.com/ros2-rust/ros2_rust/issues/217